wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/validator/DevfileIntegrityValidator.java

```diff
@@ -149,4 +200,75 @@ private void validateProjects(Devfile devfile) throws DevfileFormatException {
       }
     }
   }
+
+  private List<HasMetadata> validateSelector(Tool tool, FileContentProvider contentProvider)
+      throws ValidationException, InfrastructureException, IOException, DevfileException {
+
+    List<HasMetadata> content = getReferencedKubernetesList(tool, contentProvider);
+
+    Map<String, String> selector = tool.getSelector();
+    if (selector != null) {
+      content = SelectorFilter.filter(content, selector);
+    }
+
+    if (content.isEmpty()) {
+      throw new DevfileException(
+          format(
+              "The selector of the tool %s filters out all objects from the list.",
+              tool.getName()));
+    }
+
+    return content;
+  }
+
+  private void validateEntrypointSelector(Tool tool, List<HasMetadata> filteredObjects)
+      throws DevfileFormatException {
+
+    if (tool.getEntrypoints() == null || tool.getEntrypoints().isEmpty()) {
+      return;
+    }
+
+    for (Entrypoint ep : tool.getEntrypoints()) {
+      ContainerSearch search =
+          new ContainerSearch(
+              ep.getDeployment(),
+              ep.getPod(),
+              ep.getContainer(),
+              ep.getPodSelector(),
+              ep.getDeploymentSelector());
+
+      List<Container> cs = search.search(filteredObjects);
+
+      if (cs.isEmpty()) {
+        throw new DevfileFormatException(
+            format(
+                "Tool %s contains an entry point that doesn't match any container: %s",
+                tool.getName(), toYAML(ep)));
+      }
+    }
+  }
+
+  private List<HasMetadata> getReferencedKubernetesList(
+      Tool tool, FileContentProvider contentProvider)
+      throws ValidationException, InfrastructureException, IOException, DevfileException {
+    List<HasMetadata> content;
+    if (tool.getLocalContent() != null) {
+      content = kubernetesRecipeParser.parse(tool.getLocalContent());
+    } else if (tool.getLocal() != null) {
+      String data = contentProvider.fetchContent(tool.getLocal());
+      content = kubernetesRecipeParser.parse(data);
+    } else {
+      content = Collections.emptyList();
+    }
+
+    return content;
+  }
+
+  private static String toYAML(Entrypoint ep) {
+    try {
+      return new YAMLMapper().writer().writeValueAsString(ep);
+    } catch (JsonProcessingException e) {
+      throw new IllegalStateException("Could not serialize entrypoint to YAML. This is a bug.", e);
```

Comment:
I would avoid throwing RuntimeException here. Maybe it is better to wrap it to DevfileException?